### ISSUE TITLE
🛠 Roadmap / To Do

### ISSUE THREAD

**Issue by tannernelson** [qutheory/vapor#181]
This is Vapor's official to do list. 

We welcome anyone and everyone to help us check off all of the items on this list. If you have any questions, don't hesitate to join us on [slack.qutheory.io](http://slack.qutheory.io) to ask.

To claim a to do item, create an issue for that item on the Vapor repository (the New Issue button up on the top right), assign the issue to yourself, and explain how you are going to solve it. Then, fork the repo and get started!

- [x] SSL (using OpenSSL)
- [x] Validation
- [x] Cache
- [x] Events
- [x] File uploads
- [x] Pass [Web Frameworks Benchmark](https://www.techempower.com/benchmarks/#section=code&hw=peak&test=fortune)
- [x] Fluent
- [x] Internationalization (i18n) support
- [x] Type-safe routing
- [x] NSURLSession / cURL / Client
- [x] Foundation free (for Swift 2.2 branch)
- [x] Type-safe controllers
- [x] CRON jobs or Queue
- [x] Add QA style questions to Stack Overflow under the Vapor tag
- [x] Benchmark Vapor vs other frameworks
- [x] Look into combining Vapor with JS front-end frameworks (Meteor.js, Vue.js, Angular.js, Ember.js)
- [x] Fluent Mongo
- [x] Get the Kitura Server example working again
- [x] Add performance regression testing
- [x] Make a nice 404 page for all Vapor projects
- [x] Move the Request Date functionality into a fully Date class (to make up for lack of NSDate)
- [x] Make a Code of Conduct
- [x] Better Nginx integration + documentation
- [x] Add Zewo's MediaType to handle file types
- [x] Add CircleCI Support
- [x] Update Docs for Vapor 0.5
- [x] 100% Test coverage
- [x] 100% Documentation coverage
- [x] Websockets
- [x] Fetch dependencies on `vapor new ...`
- [x] Auth
- [x] Forums
- [x] Tutorials
- [x] Performance Optimizations
- [x] Faker Data
- [x] Heroku
- [x] Request Initializable
- [x] Hot reload/recompile
- [x] Vapor CLI self update
- [x] Deploy / Linux docs
- [x] Add benchmarks to website
- [x] Query & Form-Encoded parsing to spec

**Comment by ketzusaka** [comment]
We'll need to include Foundation to be able to wrap NSURLSession.

**Comment by tannernelson** [comment]
Yeah, Foundation Free is an old one. Hopefully as the Swift 3 release comes closer we can start using Foundation more reliably.

**Comment by mpclarkson** [comment]
@tannernelson What do you mean by Add CircleCI support? Do mean moving the testing from Travis to CircleCI? Reason for asking is that I use CircleCI at work and I know it pretty well. What are you looking for?

**Comment by shnhrrsn** [comment]
@mpclarkson Yep that's exactly it... Travis is painfully slow and CircleCI’s FAQs seem to indicate they'll set us up for free with linux/mac since we're open source

**Comment by mpclarkson** [comment]
Yes they do... they also provide 3 containers for open source projects so if you can run the tests in parallel, with some config. I'll have a go at this over the weekend... really want to get involved again but the last few weeks have been hell at work so I haven't had any time.

**Comment by tannernelson** [comment]
We have Circle CI available. Just need to get the .yml setup.

**Comment by mpclarkson** [comment]
@tannernelson I'm working on the `circle.yml` now.

**Comment by czechboy0** [comment]
Support for the Jade templating language would also be wicked.

**Comment by czechboy0** [comment]
Also a Markdown renderer would be 👍

**Comment by jchannon** [comment]
Plans for Content Negotiation?

**Comment by tannernelson** [comment]
Easier to manage these as separate issues. Let's do this in the future.
